### PR TITLE
Yakovlev/wasm experimental

### PR DESCRIPTION
**Pull request by igoriakovlev**

### CODE REVIEW

File: skiko/src/wasmTest/resources/load-test.mjs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+const karmaLoaded = window.__karma__.loaded.bind(window.__karma__);
```
Comment (igoriakovlev) [inline skiko/src/wasmTest/resources/load-test.mjs:1]: This is quite a hack, but I dont know how to suspend karma load event.

File: skiko/build.gradle.kts [hunk 2]
```diff
@@ -16,8 +16,14 @@ plugins {
     id("de.undercouch.download") version "4.1.2"
 }
 
-val coroutinesVersion = "1.6.4"
-val atomicFuVersion = "0.18.5"
+val Project.supportWasm: Boolean
```
Comment (igoriakovlev) [inline skiko/build.gradle.kts:19]: This commit is going to be squashed if such way of separation of wasm build is ok.

File: skiko/build.gradle.kts [hunk 3]
```diff
@@ -1332,3 +1389,9 @@ rootProject.plugins.withType<org.jetbrains.kotlin.gradle.targets.js.nodejs.NodeJ
         nodeExtension.download = false
     }
 }
+
+project.tasks.withType<org.jetbrains.kotlin.gradle.dsl.KotlinJsCompile>().configureEach {
+    kotlinOptions.freeCompilerArgs += listOf(
+        "-Xwasm-enable-array-range-checks", "-Xir-dce=true"
```
Comment (eymar) [inline skiko/build.gradle.kts:1395]: Does `-Xir-dce` enables DCE even for debug builds?
Do we need it for debug builds?

File: skiko/src/commonMain/kotlin/org/jetbrains/skia/Bitmap.kt [hunk 4]
```diff
@@ -1053,57 +1053,75 @@ class Bitmap internal constructor(ptr: NativePointer) : Managed(ptr, _FinalizerH
 
 
 @ExternalSymbolName("org_jetbrains_skia_Bitmap__1nGetFinalizer")
+@ModuleImport("skia", "org_jetbrains_skia_Bitmap__1nGetFinalizer")
```
Comment (eymar) [inline skiko/src/commonMain/kotlin/org/jetbrains/skia/Bitmap.kt:1056]: Is this annotation needed only for k/wasm? or for 1.8.x k/js too?

File: skiko/src/jsWasmMain/kotlin/org/jetbrains/skiko/SkiaLayer.js.kt [hunk 5]
```diff
@@ -146,7 +146,7 @@ actual open class SkiaLayer {
 
     private fun setOnChangeScaleNotifier() {
         state?.initCanvas(desiredWidth, desiredHeight, contentScale, this.pixelGeometry)
-        window.matchMedia("(resolution: ${contentScale}dppx)").addEventListener("change", { setOnChangeScaleNotifier() }, true)
+        //window.matchMedia("(resolution: ${contentScale}dppx)").addEventListener("change", { setOnChangeScaleNotifier() }, true)
```
Comment (eymar) [inline skiko/src/jsWasmMain/kotlin/org/jetbrains/skiko/SkiaLayer.js.kt:149]: Did this line cause some issue?

File: skiko/src/jvmTest/kotlin/org/jetbrains/skiko/FrameLimiterTest.kt [hunk 6]
```diff
@@ -1,274 +1,274 @@
...
-        }
-
-        override fun scheduleResumeAfterDelay(timeMillis: Long, continuation: CancellableContinuation<Unit>) {
-            val delay = ceil(timeMillis.toDouble() / delayPrecisionMillis).toInt() * delayPrecisionMillis
-            original.scheduleResumeAfterDelay(delay, continuation)
-        }
-    }
-}
\ No newline at end of file
+//package org.jetbrains.skiko
```
Comment (eymar) [inline skiko/src/jvmTest/kotlin/org/jetbrains/skiko/FrameLimiterTest.kt:1]: these are JVM tests. Do they need to be commented?

There're 2 more:
skiko/src/jvmTest/kotlin/org/jetbrains/skiko/RendezvousBroadcastChannelTest.kt
skiko/src/jvmTest/kotlin/org/jetbrains/skiko/TaskTest.kt

File: skiko/src/wasmMain/kotlin/org/jetbrains/skia/impl/Managed.wasm.kt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package org.jetbrains.skia.impl
+
+internal external class FinalizationRegistry(cleanup: (FinalizationThunk) -> Unit) {
```
Comment (eymar) [inline skiko/src/wasmMain/kotlin/org/jetbrains/skia/impl/Managed.wasm.kt:3]: It relies on js https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry, right?

File: skiko/build.gradle.kts [hunk 8]
```diff
@@ -16,8 +16,14 @@ plugins {
...
 
-val coroutinesVersion = "1.6.4"
-val atomicFuVersion = "0.18.5"
+val Project.supportWasm: Boolean
+    get() = findProperty("skiko.wasm.enabled") == "true" || isInIdea
+
+val Project.supportJs: Boolean
+    get() = findProperty("skiko.js.enabled") == "true" || isInIdea
+
+val coroutinesVersion = if (supportWasm) "1.6.4-wasm0" else "1.6.4"
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply skiko/build.gradle.kts:25]: Yes, its already published in [wasm/experimental](https://maven.pkg.jetbrains.space/kotlin/p/wasm/experimental)

File: samples/SkiaJsSample/build.gradle.kts [hunk 9]
```diff
@@ -1,11 +1,12 @@
 plugins {
-    kotlin("multiplatform") version "1.6.10"
+    kotlin("multiplatform") version "1.8.255-SNAPSHOT"
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply samples/SkiaJsSample/build.gradle.kts:2]: We are waiting for branching of kotlin 1.8.20 build soon. I suggest to push this PR after that.

File: skiko/build.gradle.kts (hunk 3 above)
Comment (igoriakovlev) [reply skiko/build.gradle.kts:1395]: This flag here affects only for tests I suppose. It does not matter to have it or not. I can remove it.

File: skiko/src/commonMain/kotlin/org/jetbrains/skia/Bitmap.kt (hunk 4 above)
Comment (igoriakovlev) [reply skiko/src/commonMain/kotlin/org/jetbrains/skia/Bitmap.kt:1056]: This annotation needed only for k/wasm.

File: skiko/src/jvmTest/kotlin/org/jetbrains/skiko/FrameLimiterTest.kt (hunk 6 above)
Comment (igoriakovlev) [reply skiko/src/jvmTest/kotlin/org/jetbrains/skiko/FrameLimiterTest.kt:1]: Ooooops.....sorry....redundant change.

File: skiko/src/wasmMain/kotlin/org/jetbrains/skia/impl/Managed.wasm.kt (hunk 7 above)
Comment (igoriakovlev) [reply skiko/src/wasmMain/kotlin/org/jetbrains/skia/impl/Managed.wasm.kt:3]: Yes, GC for wasm objects works exactly in same way as for JS, so we can use it for wasm as well.

File: skiko/src/jsWasmMain/kotlin/org/jetbrains/skiko/SkiaLayer.js.kt (hunk 5 above)
Comment (igoriakovlev) [reply skiko/src/jsWasmMain/kotlin/org/jetbrains/skiko/SkiaLayer.js.kt:149]: No, redundant change

File: samples/SkiaMultiplatformSample/build.gradle.kts [hunk 10]
```diff
@@ -110,12 +115,28 @@ kotlin {
...
+        val jsWasmMain by creating {
             dependsOn(commonMain)
             resources.setSrcDirs(resources.srcDirs)
             resources.srcDirs(unzipTask.map { it.destinationDir })
         }
 
+        val jsMain by getting {
+            dependsOn(jsWasmMain)
+            dependencies {
+                implementation("org.jetbrains.skiko:skiko-js:$version")
```
Comment (eymar) [inline samples/SkiaMultiplatformSample/build.gradle.kts:127]: if commonMain has `implementation("org.jetbrains.skiko:skiko:$version")` than jsMain and wasmMain probably don't need to add this dependency explicitly? 
Or it doesn't work like that for k/wasm?

File: samples/SkiaMultiplatformSample/build.gradle.kts [hunk 11]
```diff
@@ -91,11 +92,15 @@ kotlin {
...
+    wasm {
+        browser()
+        binaries.executable()
+    }
+
     sourceSets {
         val commonMain by getting {
             dependencies {
                 implementation("org.jetbrains.skiko:skiko:$version")
-                implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:$coroutinesVersion")
```
Comment (eymar) [inline samples/SkiaMultiplatformSample/build.gradle.kts:98]: Is it a redundant dependency (not needed at all in this example)?

File: samples/SkiaJsSample/build.gradle.kts (hunk 9 above)
[the remark this answers is not in the document]
Comment (eymar) [reply samples/SkiaJsSample/build.gradle.kts:2]: Anyway, is there a commit in kotlin repo that I can use to build it locally and try it with skiko?

File: samples/SkiaMultiplatformSample/build.gradle.kts (hunk 10 above)
Comment (igoriakovlev) [reply samples/SkiaMultiplatformSample/build.gradle.kts:127]: Yes, we can remove whole dependencies block as well. Will do.

File: samples/SkiaMultiplatformSample/build.gradle.kts (hunk 11 above)
Comment (igoriakovlev) [reply samples/SkiaMultiplatformSample/build.gradle.kts:98]: Well, I do not see any coroutine code in this sample.

File: samples/SkiaJsSample/build.gradle.kts (hunk 9 above)
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply samples/SkiaJsSample/build.gradle.kts:2]: Sure, current kotlin master (f7940a2d463d44a0830d0157373b52b8c891e0de) is fine.

**Comment by eymar** [comment]
Merging it into `wasm-main` branch. It's going to be the main branch for wasm-related progress until we migrate skiko (master) to kotlin version with all commits/features needed for wasm (probably 1.8.20?)

**Pull request merged**